File: server/handler/template.go
package handler

import (
	"io"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"

	"kvm_console/model"
	"kvm_console/service"
	"kvm_console/taskqueue"
)

// GetTemplateList 获取模板列表
func GetTemplateList(c *gin.Context) {
	templates, err := service.ListTemplates()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "获取模板列表失败: " + err.Error(),
		})
		return
	}
	role, _ := c.Get("role")
	if role != "admin" {
		filtered := make([]service.TemplateInfo, 0, len(templates))
		for _, tpl := range templates {
			if tpl.CloneVisible && !tpl.Disabled {
				filtered = append(filtered, tpl)
			}
		}
		templates = filtered
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "ok",
		"data":    templates,
	})
}

// PrepareTemplateRequest 制作模板请求
type PrepareTemplateRequest struct {
	VMName           string `json:"vm_name" binding:"required"`
	TemplateName     string `json:"template_name" binding:"required"`
	DisplayName      string `json:"display_name"`
	Compress         bool   `json:"compress"`                     // 是否压缩模板磁盘
	TransferMode     string `json:"transfer_mode"`                // copy/move；move 固定不压缩并在成功后删除源虚拟机
	Type             string `json:"type"`                         // linux/windows/fnos
	Category         string `json:"category"`                     // 二级分类，当前用于 Linux 发行版和 Windows 版本
	RootPassword     string `json:"root_password"`                // 模板 root 密码
	TemplateUser     string `json:"template_user"`                // 模板中的普通用户名
	CloudInitMode    string `json:"cloud_init_mode,omitempty"`    // 初始化模式: "nocloud"/"configdrive"/"fnos"/"none"
	PostBootCommand  string `json:"post_boot_command,omitempty"`  // Linux 模板启动后执行的自定义命令
	PostBootBlocking bool   `json:"post_boot_blocking,omitempty"` // 启动后命令阻塞模式
}

// PrepareTemplate 制作模板（异步任务）
func PrepareTemplate(c *gin.Context) {
	var req PrepareTemplateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "参数错误: 需要 vm_name 和 template_name",
		})
		return
	}
	req.TransferMode = strings.ToLower(strings.TrimSpace(req.TransferMode))
	if req.TransferMode == "" {
		req.TransferMode = service.TemplateTransferModeCopy
	}
	if req.TransferMode != service.TemplateTransferModeCopy && req.TransferMode != service.TemplateTransferModeMove {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "磁盘处理方式仅支持 copy 或 move"})
		return
	}
	if req.Compress && req.TransferMode == service.TemplateTransferModeMove {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "移动磁盘时固定为不压缩"})
		return
	}
	if req.TransferMode == service.TemplateTransferModeMove {
		if service.IsVMLocked(req.VMName) {
			c.JSON(http.StatusForbidden, gin.H{"code": 403, "message": "该虚拟机已锁定，请先解锁后再移动制作模板"})
			return
		}
		if !requireHighRiskVerification(c, "move_vm_disk_to_template") {
			return
		}
	}

	username, _ := c.Get("username")
	task, err := taskqueue.SubmitWithStruct(model.TaskTypePrepare, req, username.(string))
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "提交任务失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "模板制作任务已提交",
		"data": gin.H{
			"task_id": task.ID,
		},
	})
}

// GetLinuxTemplatePrepareCheck 获取 Linux 模板预处理前的链式依赖检查结果。
func GetLinuxTemplatePrepareCheck(c *gin.Context) {
	name := strings.TrimSpace(c.Param("name"))
	if name == "" {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "缺少模板名称"})
		return
	}
	if err := service.ValidateTemplateName(name); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	check, err := service.GetLinuxTemplatePrepareCheck(name)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"code": 500, "message": "检查模板链式依赖失败: " + err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"code": 200, "message": "ok", "data": check})
}

// PrepareImportedLinuxTemplate 为历史导入的 Linux 模板补齐离线克隆依赖。
func PrepareImportedLinuxTemplate(c *gin.Context) {
	name := strings.TrimSpace(c.Param("name"))
	if name == "" {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": "缺少模板名称"})
		return
	}
	if err := service.ValidateTemplateName(name); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"code": 400, "message": err.Error()})
		return
	}
	check, err := service.GetLinuxTemplatePrepareCheck(name)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"code": 500, "message": "检查模板链式依赖失败: " + err.Error()})
		return
	}
	if !check.CanPrepare {
		c.JSON(http.StatusConflict, gin.H{
			"code":    http.StatusConflict,
			"message": "模板链路仍有关联的链式克隆虚拟机，请先在虚拟机管理的“更多”菜单中逐台转为独立虚拟机后再预处理",
			"data":    check,
		})
		return
	}
	username, _ := c.Get("username")
	task, err := taskqueue.SubmitWithStruct(model.TaskTypeTemplateLinuxPrepare, struct {
		TemplateName string `json:"template_name"`
	}{TemplateName: name}, username.(string))
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"code": 500, "message": "提交模板预处理任务失败: " + err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "Linux 模板离线克隆依赖预处理任务已提交",
		"data":    gin.H{"task_id": task.ID},
	})
}

// GetTemplateVMs 获取模板关联虚拟机列表
func GetTemplateVMs(c *gin.Context) {
	name := c.Param("name")
	if name == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "缺少模板名称",
		})
		return
	}

	vms, err := service.ListTemplateVMs(name)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "获取模板关联虚拟机失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "ok",
		"data":    vms,
	})
}

// DeleteTemplateRequest 删除模板请求
type DeleteTemplateRequest struct {
	DeleteVMs   bool     `json:"delete_vms"`
	ExpectedVMs []string `json:"expected_vms"`
	DeleteMode  string   `json:"delete_mode"`
}

// GetDeleteTemplatePreview 获取删除模板链路预览
func GetDeleteTemplatePreview(c *gin.Context) {
	name := c.Param("name")
	if name == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "缺少模板名称",
		})
		return
	}
	preview, err := service.GetDeleteTemplatePreview(name)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "获取模板删除预览失败: " + err.Error(),
		})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "ok",
		"data":    preview,
	})
}

// DeleteTemplate 删除模板
func DeleteTemplate(c *gin.Context) {
	if !requireHighRiskVerification(c, "delete_template") {
		return
	}
	name := c.Param("name")
	if name == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "缺少模板名称",
		})
		return
	}

	var req DeleteTemplateRequest
	if err := c.ShouldBindJSON(&req); err != nil && err != io.EOF {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "删除参数错误",
		})
		return
	}

	preview, err := service.GetDeleteTemplatePreview(name)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "获取模板关联虚拟机失败: " + err.Error(),
		})
		return
	}
	deleteMode := strings.TrimSpace(req.DeleteMode)
	if deleteMode == "" {
		deleteMode = service.TemplateDeleteModeCascade
	}
	if deleteMode != service.TemplateDeleteModePromote && deleteMode != service.TemplateDeleteModePromoteHot && len(preview.RelatedVMs) > 0 && !req.DeleteVMs {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "该模板链路创建的虚拟机尚未确认删除，请重新确认后再操作",
			"data":    preview,
		})
		return
	}

	username, _ := c.Get("username")
	params := &service.DeleteTemplateParams{
		TemplateName: name,
		DeleteVMs:    req.DeleteVMs,
		ExpectedVMs:  req.ExpectedVMs,
		DeleteMode:   deleteMode,
	}

	task, err := taskqueue.SubmitWithStruct(model.TaskTypeDeleteTemplate, params, username.(string))
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "提交删除模板任务失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "删除模板任务已提交",
		"data": gin.H{
			"task_id": task.ID,
		},
	})
}

// UpdateTemplateMetaRequest 更新模板展示配置请求（兼容旧路径）
type UpdateTemplateMetaRequest struct {
	AdminName           string `json:"admin_name"`
	DisplayName         string `json:"display_name"`
	CloneVisible        bool   `json:"clone_visible"`
	Disabled            bool   `json:"disabled"`
	Category            string `json:"category"`
	VCPU                int    `json:"vcpu"`
	RAM                 int    `json:"ram"`
	DiskSize            int    `json:"disk_size"`
	DiskBus             string `json:"disk_bus"`
	NicModel            string `json:"nic_model"`
	VideoModel          string `json:"video_model"`
	CPUTopologyMode     string `json:"cpu_topology_mode"`
	FirstBootRebootMode string `json:"first_boot_reboot_mode"`
	PostBootCommand     string `json:"post_boot_command"`
	PostBootBlocking    bool   `json:"post_boot_blocking"`
}

// UpdateTemplateMeta 更新模板展示配置
func UpdateTemplateMeta(c *gin.Context) {
	name := c.Param("name")
	if name == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "缺少模板名称",
		})
		return
	}

	var req UpdateTemplateMetaRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "参数错误",
		})
		return
	}

	params := &service.UpdateTemplateMetaParams{
		AdminName:           strings.TrimSpace(req.AdminName),
		DisplayName:         strings.TrimSpace(req.DisplayName),
		CloneVisible:        req.CloneVisible,
		Disabled:            req.Disabled,
		Category:            strings.TrimSpace(req.Category),
		VCPU:                req.VCPU,
		RAM:                 req.RAM,
		DiskSize:            req.DiskSize,
		DiskBus:             req.DiskBus,
		NicModel:            req.NicModel,
		VideoModel:          req.VideoModel,
		CPUTopologyMode:     req.CPUTopologyMode,
		FirstBootRebootMode: req.FirstBootRebootMode,
		PostBootCommand:     req.PostBootCommand,
		PostBootBlocking:    req.PostBootBlocking,
	}

	if err := service.UpdateTemplateMeta(name, params); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "更新模板元数据失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    200,
		"message": "模板展示配置更新成功",
	})
}

// UpdateTemplatePublish 更新模板发布展示配置
func UpdateTemplatePublish(c *gin.Context) {
	UpdateTemplateMeta(c)
}

File: web/src/api/template.ts
/**
 * 模板相关 API
 * 对应后端 /api/template 路由组（弹性云；部分管理操作仅管理员）
 */
import service from './client'
import { API_BASE_URL } from '@/config/constants'
import { useUserStore } from '@/stores/user'
import type { ApiResponse } from '@/types/api'

/** 模板默认配置（克隆时带出推荐值） */
export interface TemplateDefaultConfig {
  vcpu?: number
  ram?: number // GB
  disk_size?: number // GB
  disk_bus?: string
  nic_model?: string
  video_model?: string
  cpu_topology_mode?: string
  first_boot_reboot_mode?: string
}

/** 模板列表项（模板树节点） */
export interface TemplateItem {
  name: string
  display_name?: string
  admin_name?: string
  type?: string // linux / windows / fnos / openwrt
  category?: string
  virtual_size?: string // 如 "20 GiB"
  actual_size?: string // 实际磁盘占用
  template_user?: string
  boot_type?: string // bios / uefi
  cloud_init_mode?: string // none 表示不初始化
  default_config?: TemplateDefaultConfig
  disabled?: boolean
  clone_visible?: boolean
  level?: number // 派生层级（管理员显示缩进）
  // ---- 模板树结构字段 ----
  template_uid?: string // 模板族标识
  node_id?: string
  parent_node_id?: string
  root_node_id?: string
  is_root?: boolean
  has_children?: boolean
  children_count?: number
  direct_vm_count?: number // 直接关联 VM 数
  tree_vm_count?: number // 子树关联 VM 数
  // ---- 导出与校验状态 ----
  exported?: boolean // 是否存在导出文件
  export_path?: string // 导出文件下载路径
  hash_status?: string // ok / missing / size_mismatch
  linux_init_status?: string // ready / failed / unknown
  md5?: string
  sha256?: string
  is_default?: boolean
  has_meta?: boolean
  created_from_vm?: string
  created_at?: string
  disk_compressed?: boolean
  source_transfer_mode?: 'copy' | 'move'
  post_boot_command?: string
  post_boot_blocking?: boolean
}

/** 获取模板列表 */
export function getTemplateList() {
  return service.get<unknown, ApiResponse<TemplateItem[]>>('/template/list')
}

/** 制作模板请求参数 */
export interface PrepareTemplatePayload {
  vm_name: string
  template_name: string
  display_name: string
  type: string
  compress?: boolean
  transfer_mode?: 'copy' | 'move'
  category?: string
  cloud_init_mode?: string
  template_user?: string
  post_boot_command?: string
  post_boot_blocking?: boolean
}

/** 制作模板（从虚拟机） */
export function prepareTemplate(data: PrepareTemplatePayload) {
  return service.post<unknown, ApiResponse<{ task_id?: string }>>('/template/prepare', data)
}

/** Linux 模板离线预处理（仅管理员，补齐 cloud-init 与磁盘扩容依赖） */
export function prepareImportedLinuxTemplate(name: string) {
  return service.post<unknown, ApiResponse<{ task_id?: string }>>(
    `/template/${encodeURIComponent(name)}/prepare-linux`,
  )
}

/** Linux 模板离线预处理链式依赖检查结果 */
export interface LinuxTemplatePrepareCheck {
  template_name: string
  linked_vms: TemplateRelatedVM[]
  can_prepare: boolean
}

/** 检查 Linux 模板预处理是否存在链式克隆依赖 */
export function getLinuxTemplatePrepareCheck(name: string) {
  return service.get<unknown, ApiResponse<LinuxTemplatePrepareCheck>>(
    `/template/${encodeURIComponent(name)}/prepare-linux/check`,
  )
}

// ==================== 模板包分片上传 ====================

/** 初始化/恢复模板包上传会话（含秒传判断） */
export function templateUploadInit(data: {
  file_name: string
  total_size: number
  file_hash: string
}) {
  return service.post<unknown, ApiResponse<{
    session_key: string
    total_chunks?: number
    chunk_size?: number
    received?: number[]
    uploaded_bytes?: number
    instant?: boolean
    completed?: boolean
  }>>('/template/upload/init', data)
}

/** 上传单个分片（multipart: file, session_key, index） */
export function templateUploadChunk(formData: FormData) {
  return service.post<unknown, ApiResponse>('/template/upload/chunk', formData, {
    timeout: 0, // 大分片上传不超时
  })
}

/** 全部分片到齐后完成校验，返回临时路径供 preview/confirm 使用 */
export function templateUploadComplete(data: { session_key: string; file_hash: string }) {
  return service.post<unknown, ApiResponse<{
    completed?: boolean
    missing?: number[]
    session_key?: string
  }>>('/template/upload/complete', data)
}

/** 清理已上传的模板包临时文件（预览后未导入而关闭对话框时调用） */
export function templateUploadCancel(path: string) {
  return service.delete<unknown, ApiResponse>('/template/upload', {
    params: { path },
  })
}

// ==================== 模板包导入 ====================

/** 导入预览节点 */
export interface ImportPreviewNode {
  name: string
  admin_name?: string
  display_name?: string
  category?: string
  template_uid?: string
  node_id?: string
  parent_node_id?: string
  root_node_id?: string
  type?: string
  clone_visible?: boolean
  disabled?: boolean
  file_size?: number
  md5?: string
  sha256?: string
  exists?: boolean
  will_import?: boolean
  conflict_reason?: string
}

/** 导入预览结果 */
export interface ImportTemplatePreview {
  token: string
  mode?: string // create=新模板树 / update=增量更新
  template_uid?: string
  root_node_id?: string
  nodes?: ImportPreviewNode[]
  can_import?: boolean
  message?: string
}

/** 解析预览模板包（source_path 为分片上传产出的临时路径或主机绝对路径） */
export function previewImportTemplate(formData: FormData) {
  return service.post<unknown, ApiResponse<{ preview?: ImportTemplatePreview; file?: string }>>(
    '/template/import/preview',
    formData,
    { timeout: 0 }, // 大文件解析不超时
  )
}

/** 确认模板包导入（提交异步任务） */
export function confirmImportTemplate(token: string) {
  return service.post<unknown, ApiResponse<{ task_id?: string }>>('/template/import/confirm', {
    token,
  })
}

// ==================== 模板导出 ====================

/** 导出模板（scope=node 导出节点 / scope=root 导出整树，提交异步任务） */
export function exportTemplate(name: string, scope: 'node' | 'root' = 'node') {
  return service.post<unknown, ApiResponse<{ task_id?: string }>>(
    `/template/${encodeURIComponent(name)}/export`,
    undefined,
    { params: { scope } },
  )
}

/** 删除模板导出文件 */
export function deleteTemplateExport(name: string) {
  return service.delete<unknown, ApiResponse>(
    `/template/${encodeURIComponent(name)}/export`,
  )
}

/** 获取模板导出文件下载地址（附带 token 供浏览器直接打开） */
export function getTemplateExportDownloadUrl(path: string): string {
  const { token } = useUserStore.getState()
  if (!path) return ''
  const normalized = path.startsWith(API_BASE_URL) ? path.slice(API_BASE_URL.length) : path
  return `${API_BASE_URL}${normalized}?token=${encodeURIComponent(token)}`
}

// ==================== 模板删除 ====================

/** 模板关联虚拟机 */
export interface TemplateRelatedVM {
  name: string
  status?: string
  ip?: string
  template?: string
  node_id?: string
  clone_mode?: string
}

/** 删除模板预览 */
export interface DeleteTemplatePreview {
  template_name?: string
  templates?: TemplateItem[]
  related_vms?: TemplateRelatedVM[]
  parent_template?: TemplateItem
  promoted_templates?: TemplateItem[]
  rebased_vms?: TemplateRelatedVM[]
  can_promote?: boolean
  promote_blockers?: string[]
  can_promote_hot?: boolean
  promote_hot_blockers?: string[]
}

/** 获取模板删除预览 */
export function getTemplateDeletePreview(name: string) {
  return service.get<unknown, ApiResponse<DeleteTemplatePreview>>(
    `/template/${encodeURIComponent(name)}/delete-preview`,
  )
}

export type TemplateDeleteMode = 'cascade' | 'promote_children' | 'promote_children_hot'

/** 删除模板（高风险操作，428 二次验证由请求层自动处理） */
export function deleteTemplate(
  name: string,
  data: {
    delete_mode?: TemplateDeleteMode
    delete_vms?: boolean
    expected_vms?: string[]
  } = {},
) {
  return service.delete<unknown, ApiResponse<{ task_id?: string }>>(
    `/template/${encodeURIComponent(name)}`,
    { data },
  )
}

// ==================== 发布设置 ====================

/** 更新模板发布展示配置参数 */
export interface UpdateTemplatePublishPayload {
  admin_name?: string
  display_name?: string
  clone_visible?: boolean
  disabled?: boolean
  category?: string
  vcpu?: number
  ram?: number
  disk_size?: number
  disk_bus?: string
  nic_model?: string
  video_model?: string
  cpu_topology_mode?: string
  first_boot_reboot_mode?: string
  post_boot_command?: string
  post_boot_blocking?: boolean
}

/** 更新模板发布展示配置 */
export function updateTemplatePublish(name: string, data: UpdateTemplatePublishPayload) {
  return service.put<unknown, ApiResponse>(
    `/template/${encodeURIComponent(name)}/publish`,
    data,
  )
}
